actors/builtin/miner/miner_actor.go

```diff
@@ -541,88 +652,172 @@ func (a Actor) TerminateSectors(rt Runtime, params *TerminateSectorsParams) *adt
-			}
-			_, fault := faultsMap[uint64(sectorNumber)]
-			Assert(fault == (sector.DeclaredFaultEpoch != epochUndefined))
-			Assert(fault == (sector.DeclaredFaultDuration != epochUndefined))
-			if !found || fault {
-				continue // Ignore declaration for missing or already-faulted sector.
+		if !empty {
+			// Check new fault are really new.
+			contains, err := abi.BitFieldContainsAny(st.Faults, newFaults)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to intersect existing faults")
+			if contains {
+				// This could happen if attempting to declare a fault for a deadline that's already passed,
+				// detected and added to Faults above.
+				// Wait for the fault detection at proving period end, or submit again omitting deadlines that have
+				// passed.
+				// Alternatively, we could subtract the detected faults from new faults.
+				rt.Abortf(exitcode.ErrIllegalArgument, "attempted to re-declare fault")
 			}
 
-			storageWeightDescs = append(storageWeightDescs, asStorageWeightDesc(st.Info.SectorSize, sector))
+			// Add new faults to state and charge fee.
+			err = st.AddFaults(store, newFaults, periodStart)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to add faults")
 
-			sector.DeclaredFaultEpoch = effectiveEpoch
-			sector.DeclaredFaultDuration = params.Duration
-			if err = st.PutSector(store, sector); err != nil {
-				rt.Abortf(exitcode.ErrIllegalState, "failed to update sector %v: %v", sectorNumber, err)
-			}
+			// Load info for sectors.
+			declaredFaultSectors, err = st.LoadSectorInfos(store, newFaults)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load fault sectors")
+
+			// Unlock penalty for declared faults.
+			declaredPenalty, err := unlockPenalty(store, &st, currEpoch, declaredFaultSectors, pledgePenaltyForSectorDeclaredFault)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to charge fault fee")
+			penalty = big.Add(penalty, declaredPenalty)
 		}
-		return temporaryFaultFee(storageWeightDescs, params.Duration)
-	}).(abi.TokenAmount)
 
-	// Burn the fee, refund any change.
-	confirmPaymentAndRefundChange(rt, requiredFee)
-	_, code := rt.Send(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, requiredFee)
-	builtin.RequireSuccess(rt, code, "failed to burn fee")
+		// Remove faulty recoveries
+		empty, err = recoveries.IsEmpty()
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to check if bitfield was empty: %s", err)
+		}
+		if !empty {
+			err = st.RemoveRecoveries(recoveries)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to remove recoveries")
+		}
+		return nil
+	})
 
-	// Request deferred Cron invocation to update temporary fault state.
-	// TODO: cant we just lazily clean this up?
-	cronPayload := CronEventPayload{
-		EventType: CronEventTempFault,
-		Sectors:   &params.SectorNumbers,
+	// Remove power for new faulty sectors.
```

Comment:
power must be removed at the deadline and not at the time it was announced

Reply:
Let's discuss that outside this PR as a potential enhancement.